nmosregistration/couchbase_backend.py

```diff
@@ -0,0 +1,202 @@
+# Copyright 2019 British Broadcasting Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+
+from gevent import monkey
+monkey.patch_all()
+
+# following block all suffixed with `# noqa E402` - follow up
+import requests
+from requests.adapters import TimeoutSauce
+from requests import Response
+import json
+import gevent
+from six.moves.urllib.parse import urlencode
+import couchbase
+import time
+from nmoscommon.timestamp import Timestamp
+import string
+from flask import make_response
+
+legacy_key_table = {
+    '@_apiversion': 'api_version'
+}
+
+def _legacy_key_lookup(key):
+    try:
+        return legacy_key_table[key]
+    except KeyError:
+        # TODO: More proper handling of any additional keys starting '@_'?
+        #       - May just be sufficient to keep table and return otherwise, need to see other cases.
+        return key
+
+class CouchbaseInterface(object):
+    def __init__(self, cluster_address, username, password, bucket, *args, **kwargs):
+        self.cluster = couchbase.cluster.Cluster('couchbase://{}'.format(','.join(cluster_address)))
+        auth = couchbase.cluster.PasswordAuthenticator(username, password)
+        self.cluster.authenticate(auth)
+        self.registry = self.cluster.open_bucket(bucket)
+        self.bucket = bucket
+
+    class RegistryUnavailable(Exception):
+        pass
+
+    def _get_associated_node_id(self, rtype, value):
+        if rtype == 'device':
+            return value['node_id']
+        elif rtype in ['receiver', 'sender', 'source', 'flow']:
+            return self.registry.get(value['device_id']).value['node_id']
+
+    def get_health(self, rkey, port=None):
+        return self.registry.lookup_in(
+            rkey, couchbase.subdocument.get('$document.exptime', xattr=True)
+        )['$document.exptime']
+
+    def upsert(self, rtype, rkey, value, xattrs, ttl=12):
+        try:
+            upsert_result = self.registry.upsert(rkey, value, ttl=ttl)
+            r = make_response(json.dumps(value), 200)
+        except couchbase.exceptions.KeyExistsError:
+            return make_response(409)
+        if upsert_result.success:
+            write_time = Timestamp.get_time().to_nanosec()
+            time.sleep(1)
```

Comment:
Any reason for this sleep?

Reply:
I think that is some debugging detritus, gone. :+1: